File: server/datasets/controlled_benchmarks/python/multi_student_cohorts/cohort_g2/structural_31_avl_tree.py
class Node_6:
    def __init__(self, v_6):
        self.v_6 = v_6
        self.r_6 = None
        self.l_6 = None
        self.height = 1

def h_6(node):
    if not node:
        return 0
    return node.height

def b_6(node):
    if not node:
        return 0
    return h_6(node.l_6) - h_6(node.r_6)

def rr_6(y):
    x = y.l_6
    sub = x.r_6
    x.r_6 = y
    y.l_6 = sub
    y.height = 1 + max(h_6(y.l_6), h_6(y.r_6))
    x.height = 1 + max(h_6(x.l_6), h_6(x.r_6))
    return x

def rl_6(x):
    y = x.r_6
    sub = y.l_6
    y.l_6 = x
    x.r_6 = sub
    x.height = 1 + max(h_6(x.l_6), h_6(x.r_6))
    y.height = 1 + max(h_6(y.l_6), h_6(y.r_6))
    return y

def ins_6(rt_6, v_6):
    if not rt_6:
        return Node_6(v_6)
    if v_6 < rt_6.v_6:
        rt_6.l_6 = ins_6(rt_6.l_6, v_6)
    elif v_6 > rt_6.v_6:
        rt_6.r_6 = ins_6(rt_6.r_6, v_6)
    else:
        return rt_6

    rt_6.height = 1 + max(h_6(rt_6.l_6), h_6(rt_6.r_6))
    bf_6 = b_6(rt_6)

    if bf_6 > 1 and v_6 < rt_6.l_6.v_6:
        return rr_6(rt_6)
    if bf_6 < -1 and v_6 > rt_6.r_6.v_6:
        return rl_6(rt_6)
    if bf_6 > 1 and v_6 > rt_6.l_6.v_6:
        rt_6.l_6 = rl_6(rt_6.l_6)
        return rr_6(rt_6)
    if bf_6 < -1 and v_6 < rt_6.r_6.v_6:
        rt_6.r_6 = rr_6(rt_6.r_6)
        return rl_6(rt_6)

    return rt_6

File: server/datasets/controlled_benchmarks/java/multi_student_cohorts/cohort_g2/structural_31_avl_tree.java
class Node_6 {
    int k_6, h_6;
    Node_6 left, right;
    Node_6(int d) {
        k_6 = d;
        h_6 = 1;
    }
}
public class Tree_6 {
    public static void main(String[] args) {
        Tree_6 t = new Tree_6();
        Node_6 r = null;
        r = t.ins_6(r, 10);
    }
    int max(int a, int b) {
        return (a > b) ? a : b;
    }
    int h_6(Node_6 N) {
        if (N == null) return 0;
        return N.h_6;
    }
    int gb_6(Node_6 N) {
        if (N == null) return 0;
        return h_6(N.left) - h_6(N.right);
    }
    Node_6 upd_6(Node_6 n_6) {
        n_6.h_6 = 1 + max(h_6(n_6.left), h_6(n_6.right));
        return n_6;
    }
    Node_6 rr_6(Node_6 y) {
        Node_6 x = y.left;
        Node_6 T2 = x.right;
        x.right = y;
        y.left = T2;
        upd_6(y);
        upd_6(x);
        return x;
    }
    Node_6 rl_6(Node_6 x) {
        Node_6 y = x.right;
        Node_6 T2 = y.left;
        y.left = x;
        x.right = T2;
        upd_6(x);
        upd_6(y);
        return y;
    }
    Node_6 ins_6(Node_6 n_6, int k_6) {
        if (n_6 == null) return (new Node_6(k_6));
        if (k_6 < n_6.k_6) n_6.left = ins_6(n_6.left, k_6);
        else if (k_6 > n_6.k_6) n_6.right = ins_6(n_6.right, k_6);
        else return n_6;
        upd_6(n_6);
        int b_6 = gb_6(n_6);
        if (b_6 > 1 && k_6 < n_6.left.k_6) return rr_6(n_6);
        if (b_6 < -1 && k_6 > n_6.right.k_6) return rl_6(n_6);
        if (b_6 > 1 && k_6 > n_6.left.k_6) {
            n_6.left = rl_6(n_6.left);
            return rr_6(n_6);
        }
        if (b_6 < -1 && k_6 < n_6.right.k_6) {
            n_6.right = rr_6(n_6.right);
            return rl_6(n_6);
        }
        return n_6;
    }
}